---
title: "A fundamental pipeline for methylation data analysis"
author: Stefano Roncelli
date: "`r format(Sys.time(), '%d/%m/%y')`"
output: pdf_document
---
# Introduction
In this work we apply some basic concepts seen in the theory lessons and apply them on a dataset of Illymina methylation data.
The dataset is composed of 8 slides, which are samples taken from four patients affected with down sydrome (`DS`) and four healthy wyld-type subjects (`WT`).

Student NÂ° 26
Address 10633381
p-value threshold: 0.01
normalization preprocessSWAN
Mann_whitney test

Starting from scratch, we have to load the necessary packages. This is done by loading the `easypackages` library that simplifies library loading procedures.
Load raw data with minfi and create an object called RGset storing the RGChannelSet object
```{r library-loading, message = FALSE, warning = FALSE}
setwd('.')
library(easypackages)
libraries("minfi", "RColorBrewer", "gap", "gplots", "tidyverse", "magrittr", "future.apply", "data.table",  "quantsmooth", "qqman", "gridExtra")
```

We can now load the experimental data store in the folder `Input_data`. We also create the `RGset` object, which containes the intensities of the Red and Green channels.
```{r loading-rgset, message = FALSE, warning = FALSE}
# TODO: fix the variables
rm(list = ls())
targets <- read.metharray.sheet("./Input_data")
RGset <- read.metharray.exp(targets = targets)
save(RGset, file = "RGset.RData")
```

# Step 2
We create the dataframes Red and Green to store the red and green fluorescences respectively, using the functions `getGreen` and `getRed`.
```{r loading-green-red, message = FALSE, warning = FALSE}
Red <- data.frame(getRed(RGset))
Green <- data.frame(getGreen(RGset))
```

# Step 3
Fill the following table: what are the Red and Green fluorescences for the address assigned to you?
Optional: check in the manifest file if the address corresponds to a Type I or a Type II probe and, in case of Type I probe, report its color.

For the optional procedure we will create two dataframes `type_I` and `type_II`, which will be reused later on for the normalization.
```{r step-3, message = FALSE, warning = FALSE}
annotation <- RGset %>%
        getAnnotation() %>%
        data.frame() %>%
        data.table() %>%
        select(Name, chr, pos, Type, AddressA, AddressB)

annotation[AddressA== 10633381 | AddressB == 10633381]
Red[rownames(Red) == '10633381',]
Green[rownames(Green) == '10633381',]
annotation[, c("AddressA", "AddressB") := NULL]
```


We can see it's a type I infinium with the Red channel.

| Sample               | Row | Column | Red Intensity | Green Intensity | Type | Color |
|----------------------|-----|--------|---------------|-----------------|------|-------|
| 5775278051           | 1   | 1      | 1852          | 458             | I    | Red   |
| 5775278051           | 4   | 2      | 1694          | 631             | I    | Red   |
| 5775278078           | 2   | 1      | 1354          | 358             | I    | Red   |
| 5775278078           | 5   | 1      | 1091          | 396             | I    | Red   |
| 5775278078           | 5   | 2      | 1131          | 424             | I    | Red   |
| 5930514034           | 1   | 2      | 796           | 302             | I    | Red   |
| 5930514035           | 4   | 2      | 894           | 354             | I    | Red   |
| 5930514035           | 6   | 2      | 1149          | 479             | I    | Red   |

# Step 4
Create the object MSet.raw
For the creation of the MSet.raw we use the preprocessRaw function
```{r step-4}
MSet.raw <- preprocessRaw(RGset)
```

# Step 5
Perform the following quality checks and provide a brief comment to each step:
- QCplot

```{r qc}
qc <- getQC(MSet.raw)
plotQC(qc)
```

- check the intensity of negative controls using minfi

```{r strip-plot}
controlStripPlot(RGset, controls = "NEGATIVE")
```
- calculate detection pValues; for each sample, how many probes have a detection p-value higher than the threshold assigned to each student?
The function to use is detectionP, which takes as input the RGset.

```{r failed-probes, message=FALSE, warning=FALSE}
detection_p_value <- detectionP(RGset)
dim(detection_p_value)
failed_probes <- detection_p_value > 0.01
table(failed_probes)
summary(failed_probes)
```
The following table summarizes the failed positions

| Sample | Group |      Slide | Row | Col  | Failed probes (p-value > 0.01) |
|--------|-------|------------|-----|------|--------------------------------|
| 1020   | DS    | 5775278051 | 1   | 1    |                            323 |
| 1036   | DS    | 5775278051 | 4   | 2    |                            260 |
| 3038   | WT    | 5775278078 | 2   | 1    |                            312 |
| 3042   | WT    | 5775278078 | 5   | 1    |                            485 |
| 3052   | WT    | 5775278078 | 5   | 2    |                            465 |
| 1016   | DS    | 5930514034 | 1   | 2    |                            123 |
| 1029   | DS    | 5930514035 | 4   | 2    |                             60 |
| 3029   | WT    | 5930514035 | 6   | 2    |                            149 |


# Step 6
Calculate raw beta and M values and plot the densities of mean methylation values, dividing the samples in DS and WT
(suggestion: subset the beta and M values matrixes in order to retain DS or WT subjects and apply the function mean to the 2 subsets).

For the retrieval of the $\beta$ and $M$ values we use the `getBeta` and `getM` functions respectively.
```{r get-intensity, message=FALSE, warning=FALSE}
beta <- MSet.raw %>%
        getBeta() %>%
        data.frame() %>%
        rownames_to_column(var = "Name") %>%
        pivot_longer(-Name, names_to = "Slide", values_to = "RawBeta") %>%
        setDT()
m <- MSet.raw %>%
        getM() %>%
        data.frame() %>%
        rownames_to_column(var = "Name") %>%
        pivot_longer(-Name, names_to = "Slide", values_to = "RawM") %>%
        setDT()

methylData <- annotation %>%
        left_join(beta) %>%
        left_join(m)


# rm(m, beta, annotation, intensities)
head(methylData)
```
We now add a column for the Sample status, which can be either Down Syndrome (`DS`) or WildType(`WT`).
```{r step-6-b, message=FALSE, warning=FALSE}
slide_DS <- colnames(RGset[,targets$Group == "WT"])
methylData %<>%
        mutate(Group = if_else(Slide %in% paste("X", slide_DS, sep = ""), "DS", "WT"), .before = RawBeta)
head(methylData)
```

Computing the mean for $\beta$ and $M$ subsets and both groups
```{r intensity-plot, message=FALSE, warning=FALSE}
rawBetaplot <- methylData %>%
        group_by(Name, Group) %>%
        mutate(mean = mean(RawBeta, na.rm = T)) %>%
        ggplot(aes(x = mean)) +
        geom_density(aes(color = Group)) +
        xlim(0, 1)
rawMplot <-methylData %>%
        group_by(Name, Group) %>%
        mutate(mean = mean(RawM, na.rm = T)) %>%
        ggplot(aes(x = mean)) +
        geom_density(aes(color = Group))
grid.arrange(rawBetaplot, rawMplot, nrow = 1)
```

# Step 7
Normalize the data using the function assigned to each student and compare raw data and normalized data.
Produce a plot with 6 panels in which, for both raw and normalized data, you show the density plots of beta mean values according to the chemistry of the probes, the density plot of beta standard deviation values according to the chemistry of the probes and the boxplot of beta values.
Provide a short comment regarding the changes you observe.

Producing the plots
```{r normalization, message=FALSE, warning=FALSE}
# Normalization with preprocessSWAN
methylData <- RGset %>%
        preprocessSWAN() %>%
        getBeta() %>%
        data.frame() %>%
        rownames_to_column(var = "Name") %>%
        pivot_longer(-Name, values_to = "NormBeta", names_to = "Slide") %>%
        left_join(methylData, ., by = c("Name", "Slide"))
```


```{r raw-vs-norm, message=FALSE, warning=FALSE}
# Density plot for raw beta mean
rawBetaMean <- methylData %>%
        group_by(Name, Type) %>%
        mutate(mean = mean(RawBeta, na.rm = T)) %>%
        ggplot(aes(x = mean)) +
        geom_density(aes(color = Type)) +
        xlim(0, 1)

# Density plot for raw beta standard deviation
rawBetaSD <- methylData %>%
        group_by(Name, Type) %>%
        mutate(sd = sd(RawBeta, na.rm = T)) %>%
        ggplot(aes(x = sd)) +
        geom_density(aes(color = Type))

# Boxplot for raw beta
rawBetaBox <- methylData %>%
        ggplot(aes(x = Slide, y = RawBeta)) +
        geom_boxplot() +
        theme(axis.text = element_blank()) +
        xlab("Slide") +
        ylab(expression("Raw" ~ beta ~ "value"))

# Density plot for Normlized beta mean
normBetaMean <- methylData %>%
        group_by(Name, Type) %>%
        mutate(mean = mean(NormBeta, na.rm = T)) %>%
        ggplot(aes(x = mean)) +
        geom_density(aes(color = Type)) +
        xlim(0, 1)

# Density plot for Normlized beta standard deviation
normBetaSD <- methylData %>%
        group_by(Name, Type) %>%
        mutate(sd = sd(NormBeta, na.rm = T)) %>%
        ggplot(aes(x = sd)) +
        geom_density(aes(color = Type))

# Boxplot for normalized beta values
normBetaBox <- methylData %>%
        ggplot(aes(x = Slide, y = NormBeta)) +
        geom_boxplot() +
        theme(axis.text = element_blank()) +
        xlab("Slide") +
        ylab(expression("Normalized"~ beta ~"value"))

grid.arrange(rawBetaMean, rawBetaSD, rawBetaBox, normBetaMean, normBetaSD, normBetaBox, nrow = 2)

```
Step 8
Perform a PCA on the beta matrix generated in step 7. Comment the plot.

```{r pca-data, message=FALSE, warning=FALSE}
pca_results <- methylData %>%
        select(Name, Slide, NormBeta) %>%
        pivot_wider( names_from = "Slide", values_from = "NormBeta") %>%
        column_to_rownames(var = 'Name') %>%
        t() %>%
        prcomp(scale = T) %T>%
        screeplot()
```
We can see how the 8th component doesn't carry any variance.

```{r pca-plot, message=FALSE, warning=FALSE}
palette("default")
groups <- factor(targets$Group)
plot(pca_results$x[, 1], pca_results$x[, 2],
  pch = 16,
  col = groups,
  xlab = "PC1",
  ylab = "PC2"
)
legend(
  "bottomleft",
  legend = levels(groups),
  col = c(1:nlevels(groups)),
  pch = 16
)
```
Step 9
Using the matrix of normalized beta values generated in step 7, identify differentially methylated probes between group DS and group WT using the functions assigned to each student.
Note; it can take several minutes; if you encounter any problem you can run the differential methylated analysis only on a subset of probes (for example those on chromosome 1, 18 and 21)


```{r mann-whitney, message=FALSE, warning=FALSE}
mann_whitney <- function(x) {
  wilcox <- wilcox.test(x ~ groups)
  return(wilcox$p.value)
}
plan(multisession)

methylData %<>%
        select(Name, Slide, NormBeta) %>%
        pivot_wider(names_from = Slide, values_from = NormBeta) %>%
        column_to_rownames(var = "Name") %>%
        future_apply(1, mann_whitney) %>%
        data.frame() %>%
        rownames_to_column(var = "Name") %>%
        data.table() %>%
        setnames(".", "PvalRaw") %>%
        left_join(methylData, ., by = "Name")
```
Step 10
Apply multiple test correction and set a significant threshold of 0.01.
How many probes do you identify as differentially methylated considering nominal pValues?
How many after Bonferroni correction?
How many after BH correction?
```{r bh-bonf, message=FALSE, warning=FALSE}
methylData %<>%
        mutate(PvalBH = p.adjust(PvalRaw,"BH"))
methylData %<>%
        mutate(PvalBonf = p.adjust(PvalRaw,"bonferroni"))
```

```{r pvalue-plot, message=FALSE, warning=FALSE}
methylData %>%
        select(PvalRaw, PvalBH, PvalBonf) %>%
        pivot_longer(everything(), names_to = "Method", values_to = "Pval") %>%
        ggplot(aes(x = Method, y = Pval)) +
        geom_boxplot() +
        ylim(0, 1)
```
Step 11
Produce an heatmap of the top 100 differentially mehtylated probes
```{r heatmaps, message=FALSE, warning=FALSE}
plot_data<- methylData %>%
        select(Name, Slide, NormBeta, PvalRaw) %>%
        pivot_wider(names_from = Slide, values_from = NormBeta) %>%
        column_to_rownames(var = 'Name') %>%
        slice_min(PvalRaw, with_ties = F, n = 100) %>%
        select(-PvalRaw) %>%
        as.matrix()

colorbar <- c("orange", "orange", "#008080", "#008080", "#008080", "orange", "orange", "#008080")
palette <- brewer.pal(100, name = "Greys")
par(mfrow = c(1, 3))

# complete linkage
heatmap.2(plot_data,
          col = palette,
          Rowv = T,
          Colv = T,
          dendrogram = "both",
          key = T,
          ColSideColors = colorbar,
          density.info = "none",
          trace = "none",
          scale = "none",
          symm = F,
          main = "Normalized Beta values - complete linkage")

# Single linkage
heatmap.2(plot_data,
          col = palette,
          Rowv = T,
          Colv = T,
          hclustfun = function(x) hclust(x, method = 'single'),
          dendrogram = "both",
          key = T,
          ColSideColors = colorbar,
          density.info = "none",
          trace = "none",
          scale = "none",
          symm = F,
          main = "Norm Beta values (Method: single linkage)")
# averge linkage
heatmap.2(plot_data,
          col = palette,
          Rowv = T,
          Colv = T, hclustfun = function(x) hclust(x, method = 'average'),
          dendrogram = "both",
          key = T, ColSideColors = colorbar,
          density.info = "none",
          trace = "none",
          scale = "none",
          symm = F,
          main = "Norm Beta values (Method: average linkage)")

```


Step 12
Produce a volcano plot and a Manhattan plot of the results of differential methylation analysis
```{r volcano, message=FALSE, warning=FALSE}
delta <- methylData %>%
                select(Name, NormBeta, Slide, Group) %>%
                group_by(Name, Group) %>%
                summarise(mean = mean(NormBeta, na.rm = T)) %>%
                group_by(Name) %>%
                summarise(Delta = diff(mean)) %>%
                select(Delta)

logbeta <- methylData %>%
        select(Name, PvalRaw) %>%
        distinct() %>%
        select(PvalRaw) %>%
        log10() %>%
        transmute(PvalRaw = -PvalRaw)
volplot_data <- cbind(delta, logbeta)
ggplot(data = volplot_data, aes(Delta, PvalRaw)) +
geom_point()
```

```{r manhattan, message=FALSE, warning=FALSE}
manhattan_data <- methylData %>%
        select(Name = Name, PvalRaw) %>%
        distinct() %>%
        merge(annotation, by = "Name") %>%
        as.data.table() %>%
        mutate(chr = numericCHR(chr))
manhattan(manhattan_data, chr="chr", bp="pos", snp="Name", p="PvalRaw")
```


Step 13 (optional)
As DS is caused by the trisomy of chromosome 21, try also to plot the density of the methylation values of the probes mapping on chromosome 21.
```{r chr21, message=FALSE, warning=FALSE}
methylData %>%
        filter(chr == "chr21") %>%
        group_by(Name, Group) %>%
        mutate(mean = mean(NormBeta, na.rm = T)) %>%
        ggplot(aes(x = mean)) +
        geom_density(aes(color = Group)) +
        xlim(0, 1)

```
Do you see a very clear difference between the samples?
How many differentially methylated probes do you find on chromosome 21?

